src/librustc_mir/ssa_analyze.rs

```diff
@@ -0,0 +1,268 @@
+    mir: &'a Mir<'tcx>,
+    param_substs: &'tcx Substs<'tcx>,
+    callbacks: C,
+) -> BitVector {
+    let mut analyzer = LocalAnalyzer::new(tcx, mir, param_substs, callbacks);
+
+    analyzer.visit_mir(mir);
+
+    for (index, ty) in mir.local_decls.iter().map(|l| l.ty).enumerate() {
+        let ty = analyzer.monomorphize(&ty);
+        debug!("local {} has type {:?}", index, ty);
+        if !analyzer.callbacks.ty_ssa_allowed(ty) {
+            analyzer.not_ssa(mir::Local::new(index));
+        }
+    }
+
+    analyzer.non_ssa_locals
+}
+
+struct LocalAnalyzer<'a, 'tcx: 'a, C: LocalAnalyzerCallbacks<'tcx>> {
+    tcx: TyCtxt<'a, 'tcx, 'tcx>,
+    mir: &'a Mir<'tcx>,
+    param_substs: &'tcx Substs<'tcx>,
+    callbacks: C,
+    dominators: Dominators<mir::BasicBlock>,
+    non_ssa_locals: BitVector,
+    // The location of the first visited direct assignment to each
+    // local, or an invalid location (out of bounds `block` index).
+    first_assignment: IndexVec<mir::Local, Location>
+}
+
+impl<'a, 'tcx: 'a, C: LocalAnalyzerCallbacks<'tcx>> LocalAnalyzer<'a, 'tcx, C> {
+    fn new(
+        tcx: TyCtxt<'a, 'tcx, 'tcx>,
+        mir: &'a Mir<'tcx>,
+        param_substs: &'tcx Substs<'tcx>,
+        callbacks: C,
+    ) -> LocalAnalyzer<'a, 'tcx, C> {
+        let invalid_location =
+            mir::BasicBlock::new(mir.basic_blocks().len()).start_location();
+        let mut analyzer = LocalAnalyzer {
+            tcx,
+            mir,
+            param_substs,
+            callbacks,
+            dominators: mir.dominators(),
+            non_ssa_locals: BitVector::new(mir.local_decls.len()),
+            first_assignment: IndexVec::from_elem(invalid_location, &mir.local_decls)
+        };
+
+        // Arguments get assigned to by means of the function being called
+        for arg in mir.args_iter() {
+            analyzer.first_assignment[arg] = mir::START_BLOCK.start_location();
+        }
+
+        analyzer
+    }
+
+    fn first_assignment(&self, local: mir::Local) -> Option<Location> {
+        let location = self.first_assignment[local];
+        if location.block.index() < self.mir.basic_blocks().len() {
+            Some(location)
+        } else {
+            None
+        }
+    }
+
+    fn not_ssa(&mut self, local: mir::Local) {
+        debug!("marking {:?} as non-SSA", local);
+        self.non_ssa_locals.insert(local.index());
+    }
+
+    fn assign(&mut self, local: mir::Local, location: Location) {
+        if self.first_assignment(local).is_some() {
+            self.not_ssa(local);
+        } else {
+            self.first_assignment[local] = location;
+        }
+    }
+
+    fn layout_of(&self, ty: Ty<'tcx>) -> TyLayout<'tcx> {
+        self.tcx.layout_of(ty::ParamEnv::reveal_all().and(ty))
+            .unwrap_or_else(|e| match e {
+                LayoutError::SizeOverflow(_) => self.tcx.sess.fatal(&e.to_string()),
+                _ => bug!("failed to get layout for `{}`: {}", ty, e)
```

Comment:
This is not in `rustc_codegen_llvm` so ideally we would treat a failure to get the error just like some arbitrary layout that we know nothing about.